ldp/graph/common_ops.py

```diff
@@ -283,6 +306,39 @@ async def forward(
 
         return result.messages[0]
 
+    async def _call_single_and_maybe_validate(
+        self, model: LLMModel, num_retries: int, **kwargs
+    ) -> LLMResult:
+        if not self.response_validator:
+            # If a response validator is not supplied, then we should not do any retries here - leave
+            # that for LiteLLM to handle.
+            return await model.call_single(**kwargs)
+
+        # NOTE: `num_retries` also gets passed to LiteLLM, so there could a maximum of
+        # `num_retries**2` retries. TODO: consider if we should have separate parameters
+        # for LiteLLM and validation retries.
+        @tenacity.retry(
+            retry=tenacity.retry_if_exception_type(ResponseValidationError),
+            # num_retries+1 because the first call is not a retry
+            stop=tenacity.stop_after_attempt(num_retries + 1),
+            wait=tenacity.wait_fixed(1),
+        )
+        async def call_and_validate() -> LLMResult:
+            result = await model.call_single(**kwargs)
+
+            try:
+                validated = cast(Callable, self.response_validator)(result)
+                if inspect.isawaitable(validated):
```

Comment:
Note we need to use `isawaitable` not `is_coroutine_callable` to properly support `Callable[[...], X | Awaitable[X]]` here (specifically the intersection in the Venn diagram would've been missed by the latter). cc @jamesbraza

Reply:
Should we modify `is_coroutine_callable`? Also, do you have test coverage for what led you to figure out this issue?

Reply:
No test - just noticed the logic: `Callable[..., X | Awaitable[X]]` captures `def f(...) -> X`, `async def f(...) -> X` and `def f(...) -> X | Awaitable[X]`. `is_coroutine_callable` wouldn't catch the last. And I don't think that's the purpose of that function, so need to modify.